### DISCUSSION TITLE
After MDX v4 migration 'React is not defined' in templates

### DISCUSSION THREAD

**Discussion by openscript** [gatsbyjs/gatsby#36663] [Help] [answered]
Dear Gatsby community

I've followed the [migration guide](https://www.gatsbyjs.com/plugins/gatsby-plugin-mdx/#migrating-from-v3-to-v4) to upgrade to gatsby-plugin-mdx version 4. 

All templates are missing the React dependency, even though `React` isn't a real dependency of my templates. Gatsby seems to inject a wrapper which needs it.

```
import { graphql, PageProps } from 'gatsby';
import { AllGenericPagesQuery } from '../../graphql-types';
import { DefaultLayout } from '../layouts/DefaultLayout';

export default function GenericPage({ children }: PageProps<AllGenericPagesQuery>) {
  return <DefaultLayout>{children}</DefaultLayout>;
}

export const query = graphql`
  query ($id: String!) {
    mdx(id: { eq: $id }) {
      id
    }
  }
`;
```

![image](https://user-images.githubusercontent.com/1105080/191738635-57a2b834-08c3-4599-9eb3-be9dec41d4ed.png)

tsconfig.json:

```
{
  "include": ["./src/**/*", "./.gatsby/**/*", "./*.ts", "./plugins/**/*"],
  "compilerOptions": {
    "target": "ESNext",
    "module": "ESNext",
    "lib": ["dom", "ES2017", "ES2019"],
    // "allowJs": true,
    // "checkJs": true,
    "jsx": "react-jsx",
    "strict": true,
    "esModuleInterop": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "noEmit": true,
    "skipLibCheck": true,
    "moduleResolution": "node",
    "resolveJsonModule": true
  }
}
```

If I `import React from 'react';` it works, but eslint complains about the unused dependency. How to fix this properly?

**Comment by LekoArts** [comment, marked as answer]
I fixed it in https://github.com/gatsbyjs/gatsby/pull/36595 - you can either use `gatsby-plugin-mdx@next` or wait until 2022-09-27 for it to be released.

**Reply by openscript** [reply]
Thank you very much @LekoArts :smiley:

**Comment by junghyeonsu** [comment]
I still have the same problem... 
my `gatsby-plugin-mdx` version is `4.3.0`.

If i `import React from 'react';` it works too.
but i think, this problem is not solved.

### Dependencies

```json
  "dependencies": {
    "@mdx-js/mdx": "^2.1.5",
    "@mdx-js/react": "^2.1.5",
    "gatsby": "^4.23.1",
    "gatsby-plugin-mdx": "^4.3.0",
    "gatsby-plugin-mdx-frontmatter": "^0.0.4",
    "gatsby-plugin-typegen": "^3.0.0",
    "gatsby-source-filesystem": "^4.24.0",
    "react": "^18.1.0",
    "react-dom": "^18.1.0",
    "remark-gfm": "v1"
  },
```

### Title Component

![image](https://user-images.githubusercontent.com/54893898/195977595-9d133dec-ce2d-4939-a992-605865202409.png)


### Layout Component

![image](https://user-images.githubusercontent.com/54893898/195977608-93193172-12b3-4fa9-8048-ec8b11ea5001.png)


### tsconfig.json

```json
{
  "compilerOptions": {
    "target": "ESNext",
    "jsx": "preserve",
    "module": "CommonJS",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "skipLibCheck": true,
    "noEmit": true
  }
}
```

<img width="954" alt="스크린샷 2022-10-15 오후 5 42 13" src="https://user-images.githubusercontent.com/54893898/195977674-94fcc281-3d2e-412b-b790-a45197beea48.png">

**Reply by LekoArts** [reply]
You're using JSX in your title component so you need to import React. That's expected

**Reply by junghyeonsu** [reply]
lol.. it was really stupid question. Thanks x)

**Comment by mbcod3** [comment]
With this tsconfig

```
{
  "compilerOptions": {
    "target": "esnext",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": false,
    "skipLibCheck": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "commonjs",
    "moduleResolution": "node",
    "noImplicitAny": true,
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "react-jsx",
    "jsxImportSource": "@emotion/react",
   }
}
```

I dont have to import react in any page. I am using emotion library. But in mdx component template file, I need to import React now when using gatsby plugin mdx v4 without that import I get React is not defined error. In v3 I didnt have to import React in the mdx component template file all other code being the same.

![react defined](https://user-images.githubusercontent.com/58457269/197650456-60a3500b-7882-47ea-bb11-e391595f75b1.png)

**Reply by LekoArts** [reply]
You need to configure https://www.gatsbyjs.com/docs/reference/config-files/gatsby-config/#jsxruntime and https://www.gatsbyjs.com/docs/reference/config-files/gatsby-config/#jsximportsource

**Reply by mbcod3** [reply]
Thanks for your response but it didnt help. I still have to import React in component file. Since its not required using that tsconfig and even the config you mentioned, vscode remove it on save since I am using organize imports feature in vscode. 

I am using latest gatsby and gatsby plugin mdx. Even in babel-preset-gatsby I have set jsxImportSource. Basically if I dont use any jsx inside component file alongside {children} from gatsby mdx I dont have to import React. I think its something to do with v4 gatsby-plugin-mdx the way its processing component file.

![mdx](https://user-images.githubusercontent.com/58457269/197828012-037dce30-6bdd-4e33-8ae6-99aba6f87d13.png)
